kv.go

```diff
@@ -0,0 +1,143 @@
+package pluginapi
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	perrors "github.com/pkg/errors"
+)
+
+// KVService provides features to set, update and retrive key-value pairs.
+type KVService struct {
+	api plugin.API
+}
+
+// SetOption used to configure a Set() operation.
+type SetOption func(*model.PluginKVSetOptions)
+
+// SetCompareCurrentValueOption returns an option to conditionally update a key's
+// value by comparing given value to match with the currently set one.
+// this is an atomic operation.
+func SetCompareCurrentValueOption(value interface{}) SetOption {
```

Comment:
Propose small nomenclature change:
```
// SetAtomic guarantees the write will occur only when the current value of matches the given old value. A client is expected to read the old value first, then pass it back to ensure the value has not since been modified.
func SetAtomic(oldValue interface{}) SetOption {
```